zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,493 @@
+"""Tuya QuirkBuilder."""
+
+from collections.abc import Callable
+from enum import Enum
+from typing import Any, Optional
+
+from zigpy.quirks import _DEVICE_REGISTRY
+from zigpy.quirks.registry import DeviceRegistry
+from zigpy.quirks.v2 import QuirkBuilder, QuirksV2RegistryEntry
+from zigpy.quirks.v2.homeassistant import EntityPlatform, EntityType
+from zigpy.quirks.v2.homeassistant.binary_sensor import BinarySensorDeviceClass
+from zigpy.quirks.v2.homeassistant.number import NumberDeviceClass
+from zigpy.quirks.v2.homeassistant.sensor import SensorDeviceClass, SensorStateClass
+import zigpy.types as t
+from zigpy.zcl import foundation
+from zigpy.zcl.clusters.measurement import (
+    RelativeHumidity,
+    SoilMoisture,
+    TemperatureMeasurement,
+)
+from zigpy.zcl.clusters.smartenergy import Metering
+
+from zhaquirks.tuya import (
+    TUYA_CLUSTER_ID,
+    PowerConfiguration,
+    TuyaLocalCluster,
+    TuyaPowerConfigurationCluster2AAA,
+)
+from zhaquirks.tuya.mcu import DPToAttributeMapping, TuyaMCUCluster, TuyaOnOffNM
+
+
+class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
+    """Tuya local RelativeHumidity cluster."""
+
+
+class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
+    """Tuya local TemperatureMeasurement cluster."""
+
+
+class TuyaSoilMoisture(SoilMoisture, TuyaLocalCluster):
+    """Tuya local SoilMoisture cluster with a device RH_MULTIPLIER factor if required."""
+
+
+class TuyaValveWaterConsumed(Metering, TuyaLocalCluster):
+    """Tuya Valve Water consumed cluster."""
+
+    VOLUME_LITERS = 0x0007
+    WATER_METERING = 0x02
+
+    """Setting unit of measurement."""
+    _CONSTANT_ATTRIBUTES = {
+        Metering.AttributeDefs.unit_of_measure.id: VOLUME_LITERS,
+        Metering.AttributeDefs.metering_device_type.id: WATER_METERING,
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Init a TuyaValveWaterConsumed cluster."""
+        super().__init__(*args, **kwargs)
+        self.add_unsupported_attribute(Metering.AttributeDefs.instantaneous_demand.id)
```

Comment:
The `LocalDataCluster` should return this as unsupported by default, right? (and thus prevent entity creation for that)
I mean, it doesn't hurt to explicitly add it as unsupported again, but is there a reason I'm missing?

Reply:
Maybe this is a bug upstream? If you don't add it to unsupported attributes, HA will create the entity.

Reply:
Hmm, I'll try and check on this in the future. The attribute is in the `REPORT_CONFIG` for the smart energy cluster, so it should be read upon pairing: https://github.com/zigpy/zha/blob/3d34a30499a80ad1802fbc72d36ad5cd5dc89ac0/zha/zigbee/cluster_handlers/smartenergy.py#L80-L84
When the attribute is read from the `LocalDataCluster`, it *should*  return "unsupported attribute", as there's no value in the attribute cache for it.
Let's leave this in then.